src/coreclr/jit/hwintrinsicxarch.cpp

```diff
@@ -2933,6 +2962,555 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
+                            // and append op3 as gtUnusedVal
+                            //
+                            // This gives us:
+                            // tmp1 = op1
+                            // tmp2 = op2
+                            // unused(op3)
+                            // res  = BinOp(tmp1, tmp2)
+
+                            spillOp1 = true;
+                            spillOp2 = true;
+
+                            std::swap(val1, val3); // 3, 2, 1
+                            std::swap(val2, val3); // 3, 1, 2
+
+                            unusedVal1 = true;
+                            break;
+                        }
+
+                        case TernaryLogicUseFlags::AC:
+                        {
+                            // We're using op1 and op3, so we need
+                            // to swap from  '1, 2, 3' to '2, 1, 3',
+                            // this means we need to spill op1 and
+                            // append op2 as gtUnusedVal
+                            //
+                            // This gives us:
+                            // tmp1 = op1
+                            // unused(op2)
+                            // res  = BinOp(tmp1, op3)
+
+                            spillOp1 = true;
+
+                            std::swap(val1, val2); // 2, 1, 3
+
+                            unusedVal1 = true;
+                            break;
+                        }
+
+                        case TernaryLogicUseFlags::BC:
+                        {
+                            // We're using op2 and op3, so we don't
+                            // need to swap, but we do need to
+                            // append op1 as gtUnusedVal
+                            //
+                            // This gives us:
+                            // * unused(op1)
+                            // * res = BinOp(op2, op3)
+
+                            unusedVal1 = true;
+                            break;
+                        }
+
+                        case TernaryLogicUseFlags::None:
+                        {
+                            // We're not using any operands, so we don't
+                            // need to swap, but we do need push all three
+                            // operands up as gtUnusedVal
+
+                            unusedVal1 = true;
+                            unusedVal2 = true;
+                            unusedVal3 = true;
+                            break;
+                        }
+
+                        default:
+                        {
+                            unreached();
+                        }
+                    }
+
+                    if (spillOp1)
+                    {
+                        impSpillSideEffect(true, verCurrentState.esStackDepth -
+                                                     3 DEBUGARG("Spilling op1 side effects for HWIntrinsic"));
+                    }
```

Comment:
This may be fine in the importer, but an example why I am leery about this complex logic: if `op1` is something that involves a local but does not get spilled by `impSpillSideEffect`, then swapping its order with `op2` and `op3` is not necessarily safe since `op2` and `op3` could assign to the local.

I _think_ it's fine as we probably don't see embedded assignments to anything that could be on the stack.

Reply:
Should I be using a different method?

`impSpillSideEffects` doesn't look for this type of interference either, so it would presumably be a bigger problem. Instead, it rather spills "special side effects" (catch args) and otherwise simply calls `impSpillSideEffect` for each stack entry up to the specified depth.

In such a scenario where `op1` is not side effecting, but `op2` and `op3` are then we'd still end up with the same logic where `op2`/`op3` get their evaluations done and appended to the statement list and corresponding local replacing the original stack entry. So we'd still end up with `op2` evaluating before `op1`.

Reply:
I don't think there is a general issue; `impAppendStmt` has a pretty clear comment inside it that it only supports the top-level assignment case and that callers are expected to spill manually for embedded cases. This was mainly just to point out that getting this kind of reordering right is pretty tricky, and only consulting effects on one of the nodes is not necessarily sufficient in the general case. But I think we can leave it like this.